### PR TITLE
fix(traces): identify artifacts using both deployed and creation code

### LINKED ISSUE #8995: bug(forge script): Failed to decode constructor arguments: buffer overrun while deserializing
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (3ff3d05 2024-10-01T00:27:45.742842000Z)

### What command(s) is the bug in?

forge script

### Operating System

macOS (Apple Silicon)

### Describe the bug

When I try to deploy the contract using the `forge script`, it crashes with a deserialization error. The error suggests a buffer overrun during the decoding of the constructor arguments.

### Script

The script I ran:

```bash
forge script scripts/DeployGlobals.s.sol \
  --sig "run()" \
  --broadcast \
  --rpc-url $RPC_URL \
  --private-key $PRIVATE_KEY
```
### Return

```
Script ran successfully.

== Return ==
globals_: contract IsleGlobals 0x0e7349cbcbA4d5394E51ae8B227a0f7D5ae3333c
2024-10-01T04:51:08.214841Z ERROR forge_script::transaction: Failed to decode constructor arguments contract=Some("ERC1967Proxy") signature=constructor(address,bytes) is_create2=false constructor_args=65676174652063616c6c206661696c65640000000000000000000000000b2bdd04d12f4fc7d4a45100ce3dc10605b44b0000000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000024c4d66de8000000000000000000000000516abfc478658666950cd371685a4fc6ce73c6c700000000000000000000000000000000000000000000000000000000
Error:
buffer overrun while deserializing

Context:
- buffer overrun while deserializing
```

Here is the [source code of my script](https://github.com/isle-labs/isle-contract/blob/fix-audit-issue/scripts/DeployGlobals.s.sol).

### PR DESCRIPTION
**Pull request by grandizzy**
## Motivation
Closes #8995 
see https://github.com/foundry-rs/foundry/pull/8998#issuecomment-2385710950

## Solution

- compute score from both runtime and creation bytecode
- minimal repro https://github.com/grandizzy/foundry-repro-8995 but not quite easy way to include in CI as depends on openzeppelin contracts lib

### CODE REVIEW

**Comment by zerosnacks** [comment]
Could you expand on how the bug raised in https://github.com/foundry-rs/foundry/issues/8995 is connected to the suggested change?

**Comment by grandizzy** [comment]
> Could you expand on how the bug raised in #8995 is connected to the suggested change? As in, how does it solve it?

ah, yep, here's the explanation https://github.com/foundry-rs/foundry/pull/8998#issuecomment-2385710950

**Review by zerosnacks — APPROVED** [review]
Ah I see, makes sense!

LGTM

File: crates/evm/traces/src/identifier/local.rs [hunk 1]
```diff
@@ -32,17 +32,28 @@ impl<'a> LocalTraceIdentifier<'a> {
...
             let contract = self.known_contracts.get(id)?;
             if let Some(deployed_bytecode) = contract.deployed_bytecode() {
-                let score = bytecode_diff_score(deployed_bytecode, code);
+                let mut score = bytecode_diff_score(deployed_bytecode, runtime_code);
+
+                // Compute and use creation code score.
+                if let Some(bytecode) = contract.bytecode() {
+                    score += bytecode_diff_score(bytecode, creation_code);
+                    score /= 2.0;
+                }
```
Comment (klkvr) [inline crates/evm/traces/src/identifier/local.rs:55]: do we actually need to check both here? are there cases when comparing just creation code is not enough?

Comment (klkvr) [reply crates/evm/traces/src/identifier/local.rs:55]: this looks like an edge case we've discussed earlier. constructor args are of length 192 while bytecode length is 1320. on >10% diff we consider bytecodes completely different.

what we could do is to only fallback to deployed bytecode comparison if we haven't found a match through comparing creation codes and/or target for exact match when comparing creation codes (i.e if we have an artifact with creation code that exactly matches the beginning of some creation bytecode in traces, then this is a 100% match)

Comment (klkvr) [reply crates/evm/traces/src/identifier/local.rs]: nice! yep, this is waht we want. could we refactor this a bit and still have a single `check` closure which would just accept a `is_creation: bool`? looks like we

Comment (grandizzy) [reply crates/evm/traces/src/identifier/local.rs]: updated in https://github.com/foundry-rs/foundry/pull/9050/commits/781b892c1037d36c98d7bb8bff6eaf1b5d8a250c I also added a test for scenario mentioned above, ptal (unrelated tests failing due to avax rpc url) Thanks!

**Review by klkvr — APPROVED** [review]
nice!

**Pull request merged**